sql/migrate/migrate.go

```diff
@@ -209,11 +262,156 @@ func DisableChecksum() PlannerOption {
 	}
 }
 
+// Plan calculates the migration Plan required for moving the current state (from) state to
+// the next state (to). A StateReader can be a directory, static schema elements or a Driver connection.
+func (p *Planner) Plan(ctx context.Context, name string, to StateReader) (*Plan, error) {
+	current, err := p.dsr.ReadState(ctx)
+	if err != nil {
+		return nil, err
+	}
+	desired, err := to.ReadState(ctx)
+	if err != nil {
+		return nil, err
+	}
+	changes, err := p.drv.RealmDiff(current, desired)
+	if err != nil {
+		return nil, err
+	}
+	if len(changes) == 0 {
+		return nil, ErrNoPlan
+	}
+	return p.drv.PlanChanges(ctx, name, changes)
+}
+
+// WritePlan writes the given Plan to the Dir based on the configured Formatter.
+func (p *Planner) WritePlan(plan *Plan) error {
+	files, err := p.fmt.Format(plan)
+	if err != nil {
+		return err
+	}
+	for _, f := range files {
+		d, err := io.ReadAll(f)
+		if err != nil {
+			return err
+		}
+		if err := p.dir.WriteFile(f.Name(), d); err != nil {
+			return err
+		}
+	}
+	if p.sum {
+		sum, err := HashSum(p.dir)
+		if err != nil {
+			return err
+		}
+		return WriteSumFile(p.dir, sum)
+	}
+	return nil
+}
+
+// NewExecutor creates a new Executor with default values. // TODO(masseelch): Operator Version and other Meta
+func NewExecutor(drv Driver, dir Dir, opts ...ExecutorOption) (*Executor, error) {
+	p := &Executor{drv: drv, dir: dir}
+	for _, opt := range opts {
+		if err := opt(p); err != nil {
+			return nil, err
+		}
+	}
+	return p, nil
+}
+
+// Execute executes n missing migration files on the database
+func (e *Executor) Execute(ctx context.Context, n int) error {
```

Comment:
In the case of `migrate apply`, I'd expect that Atlas will apply all migrations. Do we want a custom method for this, or do you want to accept -1 as a parameter to indicate "all"?

Reply:
My plan is to accept `-1` as this is common in other implementations (e.g. `strings.SplitN()`